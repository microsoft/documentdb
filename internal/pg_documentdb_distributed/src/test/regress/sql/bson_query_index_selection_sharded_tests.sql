
SET search_path TO helio_core,helio_api,helio_api_catalog,helio_api_internal;
SET citus.next_shard_id TO 3610000;
SET helio_api.next_collection_id TO 3610;
SET helio_api.next_collection_index_id TO 3610;

-- insert 10K documents
SELECT COUNT (*) FROM ( SELECT helio_api.insert_one('db', 'test_index_selection_sharded', FORMAT('{ "a": { "b": %s, "c": %s } }', i, i)::bson) FROM generate_series(1, 10000) i) r1;

SELECT helio_distributed_test_helpers.drop_primary_key('db', 'test_index_selection_sharded');

-- create indexes on a.b, and a.c
SELECT helio_api_internal.create_indexes_non_concurrently('db', '{ "createIndexes": "test_index_selection_sharded", "indexes": [ { "name": "a_b_1", "key": { "a.b": 1 } }, { "name": "a_c_1", "key": { "a.c": 1 }}] }', true);

-- Now, do an explain with an OR query that each uses 1 of the indexes.
EXPLAIN (VERBOSE ON, COSTS OFF) SELECT document FROM helio_api.collection('db', 'test_index_selection_sharded') WHERE document @@ '{ "$or": [ { "a.b": { "$gt": 500 } }, { "a.c": { "$lt": 10 } } ] }';

EXPLAIN (VERBOSE ON, COSTS OFF) SELECT document FROM helio_api.collection('db', 'test_index_selection_sharded') WHERE document @@ '{ "$and": [ { "a.b": { "$gt": 500 } }, { "a.c": { "$lt": 10 } } ] }';

BEGIN;
set local citus.enable_local_execution to off;
EXPLAIN (VERBOSE ON, COSTS OFF) SELECT document FROM helio_api.collection('db', 'test_index_selection_sharded') WHERE document @@ '{ "$or": [ { "a.b": { "$gt": 500 } }, { "a.c": { "$lt": 10 } } ] }';
ROLLBACK;


-- Now shard the collection
SELECT helio_api.shard_collection('db', 'test_index_selection_sharded', '{ "_id": "hashed" }', false);

-- rerun the query
BEGIN;
set local enable_seqscan to off;
EXPLAIN (VERBOSE ON, COSTS OFF) SELECT document FROM helio_api.collection('db', 'test_index_selection_sharded') WHERE document @@ '{ "$or": [ { "a.b": { "$gt": 500 } }, { "a.c": { "$lt": 10 } } ] }';

SET LOCAL helio_api.ForceUseIndexIfAvailable to OFF;
EXPLAIN (VERBOSE ON, COSTS OFF) SELECT document FROM helio_api.collection('db', 'test_index_selection_sharded') WHERE document @@ '{ "$and": [ { "a.b": { "$gt": 500 } }, { "a.c": { "$lt": 10 } } ] }';

ROLLBACK;
