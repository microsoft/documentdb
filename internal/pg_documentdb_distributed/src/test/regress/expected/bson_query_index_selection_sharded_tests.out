SET search_path TO helio_core,helio_api,helio_api_catalog,helio_api_internal;
SET citus.next_shard_id TO 3610000;
SET helio_api.next_collection_id TO 3610;
SET helio_api.next_collection_index_id TO 3610;
-- insert 10K documents
SELECT COUNT (*) FROM ( SELECT helio_api.insert_one('db', 'test_index_selection_sharded', FORMAT('{ "a": { "b": %s, "c": %s } }', i, i)::bson) FROM generate_series(1, 10000) i) r1;
NOTICE:  creating collection
 count 
---------------------------------------------------------------------
 10000
(1 row)

SELECT helio_distributed_test_helpers.drop_primary_key('db', 'test_index_selection_sharded');
 drop_primary_key 
---------------------------------------------------------------------
 
(1 row)

-- create indexes on a.b, and a.c
SELECT helio_api_internal.create_indexes_non_concurrently('db', '{ "createIndexes": "test_index_selection_sharded", "indexes": [ { "name": "a_b_1", "key": { "a.b": 1 } }, { "name": "a_c_1", "key": { "a.c": 1 }}] }', true);
                                                                                                   create_indexes_non_concurrently                                                                                                    
---------------------------------------------------------------------
 { "raw" : { "defaultShard" : { "numIndexesBefore" : { "$numberInt" : "0" }, "numIndexesAfter" : { "$numberInt" : "2" }, "createdCollectionAutomatically" : false, "ok" : { "$numberInt" : "1" } } }, "ok" : { "$numberInt" : "1" } }
(1 row)

-- Now, do an explain with an OR query that each uses 1 of the indexes.
EXPLAIN (VERBOSE ON, COSTS OFF) SELECT document FROM helio_api.collection('db', 'test_index_selection_sharded') WHERE document @@ '{ "$or": [ { "a.b": { "$gt": 500 } }, { "a.c": { "$lt": 10 } } ] }';
                                                                                                                                                                               QUERY PLAN                                                                                                                                                                               
---------------------------------------------------------------------
 Custom Scan (Citus Adaptive)
   Output: remote_scan.document
   Task Count: 1
   Tasks Shown: All
   ->  Task
         Query: SELECT document FROM helio_data.documents_3610_3610002 collection WHERE (((document OPERATOR(helio_api_catalog.#>) '{ "a.b" : { "$numberInt" : "500" } }'::helio_core.bsonquery) OR (document OPERATOR(helio_api_catalog.#<) '{ "a.c" : { "$numberInt" : "10" } }'::helio_core.bsonquery)) AND (shard_key_value OPERATOR(pg_catalog.=) '3610'::bigint))
         Node: host=localhost port=58070 dbname=regression
         ->  Bitmap Heap Scan on helio_data.documents_3610_3610002 collection
               Output: document
               Recheck Cond: ((collection.document OPERATOR(helio_api_catalog.@>) '{ "a.b" : { "$numberInt" : "500" } }'::helio_core.bson) OR (collection.document OPERATOR(helio_api_catalog.@<) '{ "a.c" : { "$numberInt" : "10" } }'::helio_core.bson))
               ->  BitmapOr
                     ->  Bitmap Index Scan on a_b_1
                           Index Cond: (collection.document OPERATOR(helio_api_catalog.@>) '{ "a.b" : { "$numberInt" : "500" } }'::helio_core.bson)
                     ->  Bitmap Index Scan on a_c_1
                           Index Cond: (collection.document OPERATOR(helio_api_catalog.@<) '{ "a.c" : { "$numberInt" : "10" } }'::helio_core.bson)
(15 rows)

EXPLAIN (VERBOSE ON, COSTS OFF) SELECT document FROM helio_api.collection('db', 'test_index_selection_sharded') WHERE document @@ '{ "$and": [ { "a.b": { "$gt": 500 } }, { "a.c": { "$lt": 10 } } ] }';
                                                                                                                                                                              QUERY PLAN                                                                                                                                                                               
---------------------------------------------------------------------
 Custom Scan (Citus Adaptive)
   Output: remote_scan.document
   Task Count: 1
   Tasks Shown: All
   ->  Task
         Query: SELECT document FROM helio_data.documents_3610_3610002 collection WHERE ((document OPERATOR(helio_api_catalog.#>) '{ "a.b" : { "$numberInt" : "500" } }'::helio_core.bsonquery) AND (document OPERATOR(helio_api_catalog.#<) '{ "a.c" : { "$numberInt" : "10" } }'::helio_core.bsonquery) AND (shard_key_value OPERATOR(pg_catalog.=) '3610'::bigint))
         Node: host=localhost port=58070 dbname=regression
         ->  Bitmap Heap Scan on helio_data.documents_3610_3610002 collection
               Output: document
               Recheck Cond: ((collection.document OPERATOR(helio_api_catalog.@<) '{ "a.c" : { "$numberInt" : "10" } }'::helio_core.bson) AND (collection.document OPERATOR(helio_api_catalog.@>) '{ "a.b" : { "$numberInt" : "500" } }'::helio_core.bson))
               ->  BitmapAnd
                     ->  Bitmap Index Scan on a_c_1
                           Index Cond: (collection.document OPERATOR(helio_api_catalog.@<) '{ "a.c" : { "$numberInt" : "10" } }'::helio_core.bson)
                     ->  Bitmap Index Scan on a_b_1
                           Index Cond: (collection.document OPERATOR(helio_api_catalog.@>) '{ "a.b" : { "$numberInt" : "500" } }'::helio_core.bson)
(15 rows)

BEGIN;
set local citus.enable_local_execution to off;
EXPLAIN (VERBOSE ON, COSTS OFF) SELECT document FROM helio_api.collection('db', 'test_index_selection_sharded') WHERE document @@ '{ "$or": [ { "a.b": { "$gt": 500 } }, { "a.c": { "$lt": 10 } } ] }';
                                                                                                                                                                               QUERY PLAN                                                                                                                                                                               
---------------------------------------------------------------------
 Custom Scan (Citus Adaptive)
   Output: remote_scan.document
   Task Count: 1
   Tasks Shown: All
   ->  Task
         Query: SELECT document FROM helio_data.documents_3610_3610002 collection WHERE (((document OPERATOR(helio_api_catalog.#>) '{ "a.b" : { "$numberInt" : "500" } }'::helio_core.bsonquery) OR (document OPERATOR(helio_api_catalog.#<) '{ "a.c" : { "$numberInt" : "10" } }'::helio_core.bsonquery)) AND (shard_key_value OPERATOR(pg_catalog.=) '3610'::bigint))
         Node: host=localhost port=58070 dbname=regression
         ->  Bitmap Heap Scan on helio_data.documents_3610_3610002 collection
               Output: document
               Recheck Cond: ((collection.document OPERATOR(helio_api_catalog.@>) '{ "a.b" : { "$numberInt" : "500" } }'::helio_core.bson) OR (collection.document OPERATOR(helio_api_catalog.@<) '{ "a.c" : { "$numberInt" : "10" } }'::helio_core.bson))
               ->  BitmapOr
                     ->  Bitmap Index Scan on a_b_1
                           Index Cond: (collection.document OPERATOR(helio_api_catalog.@>) '{ "a.b" : { "$numberInt" : "500" } }'::helio_core.bson)
                     ->  Bitmap Index Scan on a_c_1
                           Index Cond: (collection.document OPERATOR(helio_api_catalog.@<) '{ "a.c" : { "$numberInt" : "10" } }'::helio_core.bson)
(15 rows)

ROLLBACK;
-- Now shard the collection
SELECT helio_api.shard_collection('db', 'test_index_selection_sharded', '{ "_id": "hashed" }', false);
 shard_collection 
---------------------------------------------------------------------
 
(1 row)

-- rerun the query
BEGIN;
set local enable_seqscan to off;
EXPLAIN (VERBOSE ON, COSTS OFF) SELECT document FROM helio_api.collection('db', 'test_index_selection_sharded') WHERE document @@ '{ "$or": [ { "a.b": { "$gt": 500 } }, { "a.c": { "$lt": 10 } } ] }';
                                                                                                                                                  QUERY PLAN                                                                                                                                                  
---------------------------------------------------------------------
 Custom Scan (Citus Adaptive)
   Output: remote_scan.document
   Task Count: 8
   Tasks Shown: One of 8
   ->  Task
         Query: SELECT document FROM helio_data.documents_3610_3610016 documents_3610 WHERE ((document OPERATOR(helio_api_catalog.#>) '{ "a.b" : { "$numberInt" : "500" } }'::helio_core.bsonquery) OR (document OPERATOR(helio_api_catalog.#<) '{ "a.c" : { "$numberInt" : "10" } }'::helio_core.bsonquery))
         Node: host=localhost port=58070 dbname=regression
         ->  Bitmap Heap Scan on helio_data.documents_3610_3610016 documents_3610
               Output: document
               Recheck Cond: ((documents_3610.document OPERATOR(helio_api_catalog.@>) '{ "a.b" : { "$numberInt" : "500" } }'::helio_core.bson) OR (documents_3610.document OPERATOR(helio_api_catalog.@<) '{ "a.c" : { "$numberInt" : "10" } }'::helio_core.bson))
               ->  BitmapOr
                     ->  Bitmap Index Scan on a_b_1
                           Index Cond: (documents_3610.document OPERATOR(helio_api_catalog.@>) '{ "a.b" : { "$numberInt" : "500" } }'::helio_core.bson)
                     ->  Bitmap Index Scan on a_c_1
                           Index Cond: (documents_3610.document OPERATOR(helio_api_catalog.@<) '{ "a.c" : { "$numberInt" : "10" } }'::helio_core.bson)
(15 rows)

SET LOCAL helio_api.ForceUseIndexIfAvailable to OFF;
EXPLAIN (VERBOSE ON, COSTS OFF) SELECT document FROM helio_api.collection('db', 'test_index_selection_sharded') WHERE document @@ '{ "$and": [ { "a.b": { "$gt": 500 } }, { "a.c": { "$lt": 10 } } ] }';
                                                                                                                                                  QUERY PLAN                                                                                                                                                   
---------------------------------------------------------------------
 Custom Scan (Citus Adaptive)
   Output: remote_scan.document
   Task Count: 8
   Tasks Shown: One of 8
   ->  Task
         Query: SELECT document FROM helio_data.documents_3610_3610016 documents_3610 WHERE ((document OPERATOR(helio_api_catalog.#>) '{ "a.b" : { "$numberInt" : "500" } }'::helio_core.bsonquery) AND (document OPERATOR(helio_api_catalog.#<) '{ "a.c" : { "$numberInt" : "10" } }'::helio_core.bsonquery))
         Node: host=localhost port=58070 dbname=regression
         ->  Bitmap Heap Scan on helio_data.documents_3610_3610016 documents_3610
               Output: document
               Recheck Cond: (documents_3610.document OPERATOR(helio_api_catalog.@<) '{ "a.c" : { "$numberInt" : "10" } }'::helio_core.bson)
               Filter: (documents_3610.document OPERATOR(helio_api_catalog.@>) '{ "a.b" : { "$numberInt" : "500" } }'::helio_core.bson)
               ->  Bitmap Index Scan on a_c_1
                     Index Cond: (documents_3610.document OPERATOR(helio_api_catalog.@<) '{ "a.c" : { "$numberInt" : "10" } }'::helio_core.bson)
(13 rows)

ROLLBACK;
